### ISSUE TITLE
Providing __filename and __dirname "module locals"

### ISSUE THREAD

**Issue by saghul** [svaarala/duktape#754]
I've been trying to come up with a way to provide `__filename` and `__dirname` module local functions in sjs, and I'm kind of stuck.

For binary modules, since I set the exports, I could add those myself, since I know where the file was loaded from, but for pure JS modules, since we just return the source code, which could reassign module.exports, we'd lose it. There is also the case of a normal file we are evaluating, not inside any module.

Any ideas?

**Comment by saghul** [comment]
I managed to implement this in sjs: https://github.com/saghul/sjs/commit/7ac40063593f5913c6d204c0a5a13e79ad2d3f88 it consists of 2 parts:

* when executing a file, I set a global `__file__` attribute
* when loading a module using require, I prepend `var __file__ = XXX`; to the source code

It seems to work well in practice, though it would be nice if the modSearch wrapper considered it.

**Comment by fatcerberus** [comment]
You should be able to just set the `module.filename` in modSearch?  And any other properties you need on the `module` object.

**Comment by saghul** [comment]
Yeah, I thought of that, but I want to get the same usage pattern when
requiring and when running a file standalone.

**Comment by fatcerberus** [comment]
For what it's worth, I think the way Node.js handles that is by treating _everything_ as a module--even the "main" script.  Not sure what the situation is in sjs though.

**Comment by saghul** [comment]
I'm not sure they do that, but it's definitely possible! In sjs I don't do that, but I might revisit in the future.

**Comment by fatcerberus** [comment]
> I'm not sure they do that

They do, that's what `require.main` is. :)  See here:
https://nodejs.org/api/modules.html#modules_accessing_the_main_module

That said, Node *does* provide "global" (actually module-scoped) `__filename` and `__dirname` which I didn't realize.

**Comment by saghul** [comment]
Oh! Heh, this is weird:

~~~~
$  node
> module
Module {
  id: '<repl>',
  exports: {},
  parent: undefined,
  filename: null,
  loaded: false,
  children: [],
  paths:
   [ '/Users/saghul/repl/node_modules',
     '/Users/saghul/node_modules',
     '/Users/node_modules',
     '/node_modules',
     '/Users/saghul/.node_modules',
     '/Users/saghul/.node_libraries',
     '/Users/saghul/.nvm/versions/node/v4.4.4/lib/node' ] }
~~~~

Not sure if I want to emulate that model though...

**Comment by fatcerberus** [comment]
It actually serves a useful purpose: It ensures that e.g. `var` or functions at top level in the main file don't accidentally become global, while still allowing other modules to call stuff from main (through its `exports`).

**Comment by fatcerberus** [comment]
@saghul I just ended up going whole hog and implemented my own Node.js-like module loader and `require()`.  It's not difficult, just tedious to construct all the objects on the stack.  Here's my implementation if you want to take notes, minisphere is BSD licensed :)

https://github.com/fatcerberus/minisphere/blob/master/src/engine/commonjs.c

`duk_push_require_function` and `duk_peval_module` are where the magic happens.

**Comment by saghul** [comment]
Sweet, thanks Bruce! I'll check it out!

**Comment by svaarala** [comment]
@fatcerberus I could add a Wiki page of module loaders to look at and link the minisphere loader there?

**Comment by fatcerberus** [comment]
Sure, go right ahead.  I'd actually like to generalize the code a bit in the future so it can be plugged into any Duktape app.  Right now it's a bit minisphere-specific, in particular the module ID resolution and source code retrieval should be pluggable a la the modResolve/modLoad proposal.
